provider/pkg/await/deployment.go

```diff
@@ -341,6 +344,10 @@ func (dia *deploymentInitAwaiter) await(
 		}
 
 		// Else, wait for updates.
+		// We run the event handlers in goroutines to avoid blocking the main event loop.
+		// We also need to ensure that only one of each event processes can run at a time.
+		var depMutex, rsMutex, podMutex, pvcMutex sync.Mutex
```

Comment:
I'm a little worried this introduces room for more race conditions on `dia`, because we can still be handling multiple events at once. Re: not blocking the main loop, do you have an idea of where we're getting bottlenecked?

I wonder if there's an alternative where we simply ignore replicaset events until we see a deployment event with a generation change. Something like:

```go
select {
case event := <-deploymentEvents:
    deploying := dia.processDeploymentEvent(event)
case event := <-replicaSetEvents:
    if deploying {
        dia.processReplicaSetEvent(event)
    }
}
```

Reply:
Yes, the initial approach in this PR may be risky. We also shouldn't be discarding replicaset events since we need those events to inform us of status updates so I'm hesitant to go down the approach you suggested.

I've reworked this PR so that we continuously poll for deployment events first, and ensure that the deployment controller has informed us if a rollout was triggered or not before we proceed onto processing other events. This ensures that when we check the replicaset, the deployment controller would have informed us if a roll-out has been triggered.